### PR TITLE
665 666 669 unallocated line delete/update service and outbound shipment status update restriction

### LINKED ISSUE #665: Add unallocated line update service
https://github.com/openmsupply/remote-server/issues/531

### LINKED ISSUE #669: Outbound status change to allocated only allowed when there are no unallocated lines

### PR DESCRIPTION
**Pull request by andreievg**
Closes #665 
Coses #666 
Closes #669

Should be pretty straight forward, as per the issues

### CODE REVIEW

File: service/src/invoice_line/validate.rs [hunk 1]
```diff
@@ -92,6 +92,19 @@ pub fn check_line_exists(
         Err(error) => Err(WithDBError::db(error)),
     }
 }
+// TODO use this one instead of check_line_exists
+pub fn check_line_exists_option(
+    connection: &StorageConnection,
+    id: &str,
+) -> Result<Option<InvoiceLineRow>, RepositoryError> {
+    let result = InvoiceLineRowRepository::new(connection).find_one_by_id(id);
```
Comment (clemens-msupply) [inline service/src/invoice_line/validate.rs:100]: .find_one_by_id(id).optional()

File: service/src/invoice_line/outbound_shipment_unallocated_line/update.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+fn generate(
+    UpdateOutboundShipmentUnallocatedLine {
+        id: _,
+        quantity,
+    }: UpdateOutboundShipmentUnallocatedLine,
+    mut line: InvoiceLineRow,
+) -> Result<InvoiceLineRow, UpdateOutboundShipmentUnallocatedLineError> {
+    line.number_of_packs = u32_to_i32(quantity);
```
Comment (clemens-msupply) [inline service/src/invoice_line/outbound_shipment_unallocated_line/update.rs:70]: As in the other PR, in general I would  keep the input immutable and clone it instead (even though this method is very simple)

**Review by clemens-msupply — APPROVED** [review]

File: service/src/invoice_line/outbound_shipment_unallocated_line/update.rs (hunk 2 above)
Comment (andreievg) [reply service/src/invoice_line/outbound_shipment_unallocated_line/update.rs:70]: There is no conditions here, i am not sure why it needs to be immutable, it's not a parameter that is leaked out of this method.

File: service/src/invoice_line/validate.rs (hunk 1 above)
Comment (andreievg) [reply service/src/invoice_line/validate.rs:100]: as per our convo, i think optional would only work when operations are done on dsl

**Pull request merged**